Preview first mockup after updating image

YayunHuang opened the pull request:
Ref: MUP-137


https://github.com/user-attachments/assets/f467e071-2749-41f8-acf3-085a9d18614f

pkong-ds commented on public/mockup.zip:
curious: Did you run `make bundle-python` or zip manually? 😆

YayunHuang replied:
run `make bundle-python`

pkong-ds commented on mockup_package/mockup/image.py:
```diff
@@ -43,6 +43,48 @@ async def mockup(location, device_id, original_img_path_list, device_info):
 
 
+# Now only the first orientation model is generated for preview
+async def previewMockup(
+    location, device_id, original_img_path_list, device_info, preview_img_index=0
+):
```
1. should we support `preview_img_index: List[int] = [0]` for future?
2. should we add types in python from now on? seems more readable

```python
async def previewMockup(
preview_img_index: List[int] = [0]
) -> List[someType]: # or Tuple better?
```

> Starting with Python 3.6, you can declare types of variables and functions, like this :

> `explicit_number: type`

YayunHuang replied:
@pkong-ds I think we won't preview more than one orientation at once? if the user change orientation or upload new image, should generate a new preview image for it
and for types in python, I think yes, will add the type if I find the type missing

pkong-ds replied:
Which UX do you think is better? After user upload image
1. orientiationA loading -> show oriA preview -> click oriB -> _oriB loading_ -> show oriB preview
2. allOri loading -> show oriA preview -> click oriB -> show oriB preview

Or even better
3. Can we parallel this somehow, so that generation of oriA would not block oriB?

pkong-ds replied:
> if the user change orientation or upload new image, should generate a new preview image for it

btw, according to this logic, `original_img_path_list: List[str]` can just be `original_img_path: str`?

YayunHuang replied:
@pkong-ds I think 1 is better because case 2 the user needs to wait longer,, but we can store the result once orientiationA generated and don't need to generate again if user change back to select orientiationA.

`original_img_path_list: List[str]` is for the uploaded images, and `preview_image_index` is for which orientation use for preview, let me update `preview_image_index` to `preview_orientation_index` to avoid confusion

pkong-ds replied:
Thanks a lot. 

I was saying original_path because we wont process multiple images here? It's always a List of ONE single item?

That's why I suggested changing it as a str variable, over List[str]

YayunHuang replied:
@pkong-ds oh, as the logic is almost same, I update to merge `mockup` and `previewMockup` to one function now, and will generate all mockup when `preview_orientation_index = None`

pkong-ds replied:
@YayunHuang I disagree with merging 2 as one function. Please let me know what you think about below

1. they serve 2 purposes, as you mentioned, `mockup` & `previewMockup`
2. IMO `their logic being same now` != `we should merge the f unction`. Imagine if future, changing features about `previewMockup` will change this shared function, which might leads to future devs misunderstanding such changes as related to `mockup`
3. `original_img_path` list is a good example of incorrect abstraction already.
    - `mockup` supports list of images
    - `previewMockup` always expects single image
4. You can extract the `generate` part as a separate function

```js
function generate() {  
  // ... duplicate logic
}

function mockup {  
  generate()
}

function previewMockup {
  generate()
}

```

pkong-ds commented on mockup_package/mockup/image.py:
```diff
@@ -43,6 +43,48 @@ async def mockup(location, device_id, original_img_path_list, device_info):
+async def previewMockup(
+    location, device_id, original_img_path_list, device_info, preview_img_index=0
+):
+    device_path_prefix = f"{location.split('/')[0]}//{location.split('/')[2]}"
+    device_mask_path_prefix = device_path_prefix + "/images/mockup_mask_templates/"
+    device_path_prefix += "/images/mockup_templates/"
+    device_path = "./device.png"
+    device_mask_path = "./device_mask.png"
+    output_img_path_list = []
+    for original_img_path in original_img_path_list:
+        ig = IG(original_img_path, device_id, device_info)
+        ig.create_fit_resolution_image()
+        try:
+            spec = list(ig.phone_models.get(device_id).get("mockups").values())[
+                preview_img_index
+            ]
+            await process_response(
+                device_path_prefix + str(spec["image"]),
+                device_path,
+            )
+            await process_response(
+                device_mask_path_prefix + str(spec["image"]),
+                device_mask_path,
+            )
+        except Exception as e:
+            print(e, file=sys.stderr)
+            # js.errorBox(e)
+            raise
+        ig.create_fit_coord_image(spec)
+        deviceView = str(spec["image"]).split("-")[-1].split(".")[0]
```
this part seems duplicate with above. Can refactor as separate function?

pkong-ds commented on public/scripts/upload.js:
```diff
@@ -88,7 +129,9 @@ class FileListViewModel {
     for (const file of files) {
       const imageUpload = new ImageUpload(file, MAX_FILE_SIZE_BYTE);
       await imageUpload.read();
+      imageUpload.id = Date.now();
```
or generate a [ULID](https://github.com/ulid/javascript)?

pkong-ds commented on public/scripts/upload.js:
```diff
@@ -44,6 +44,47 @@ async function runWorker(worker) {
+          }
+        })
+        .catch(function (err) {
+          console.error("Get error while storing images to localforage:", err);
+        });
```
Maybe Add TODO: Handle preview error in widget

pkong-ds commented on public/scripts/preview_worker.js:
```diff
@@ -0,0 +1,66 @@
+    self["deviceInfo"] = event.data.deviceInfo;
+
+    try {
+      let results = await runPreviewMockup(pyodideObject);
```
Maybe add TODO: handle preview loading state in widget

pkong-ds commented on src/styles/upload.css:
```diff
@@ -147,6 +147,8 @@ main {
   width: 100%;
   height: 100%;
   object-fit: contain;
+  position: relative;
```
If this new relative overrides some other relative, should some other css class remove `position:relative`?

pkong-ds reviewed (commented):
The javascript and python are not very readable :sadfrog: maybe we should prioritize 
- enforce typescript
- allow code sharing in `scripts/`

pkong-ds commented on mockup_package/mockup/image.py:
```diff
@@ -1,20 +1,25 @@
-    output_img_path_list = []
+    device_path: str = "./device.png"
+    device_mask_path: str = "./device_mask.png"
+    output_img_path_list: list[tuple[str, str, str]] = []
```
Not related to this PR. Maybe we should introduce `pydantic` to do sth like

```
class OutpuImgPath:
  path
  original_img_path
  device_view

output_img_path_list: list[OutputImgPath]
```

pkong-ds commented on public/scripts/ulid.min.js:
Add version

pkong-ds replied:
test slack integration comment

YayunHuang commented on public/scripts/ulid.min.js:
```diff
@@ -0,0 +1,130 @@
+/**
+ * Minified by jsDelivr using Terser v5.19.2.
+ * Original file: /npm/ulid@2.3.0/dist/index.umd.js
```
@pkong-ds the file already contains version number

pkong-ds replied:
LOL, didn't expect a `min.js` to have spaces & indents for readability.

Coz normally dev is not expected to `open a min.js file and read it`.

Anyway, think this is good enough

YayunHuang commented:
@pkong-ds updated to extract the `generate` function which only generates one image file and one orientation(spec in code), and I also updated and implemented some functions so that we can only pass a single file instead of a list of files when generating the preview mockup, please review, thanks

pkong-ds reviewed (approved):
LGTM. Thanks a lot yayun!
